lib/config_cat/config_fetcher.ex

```diff
@@ -34,14 +43,49 @@ defmodule ConfigCat.CacheControlConfigFetcher do
 
     initial_state =
       default_options()
-      |> Keyword.merge(options)
-      |> Enum.into(%{})
+      |> custom_server_options(options)
+      |> Map.merge(Enum.into(options, %{}))
+      |> data_governance_options(options)
       |> Map.put(:etag, nil)
 
     GenServer.start_link(__MODULE__, initial_state, name: name)
   end
 
-  defp default_options, do: [api: ConfigCat.API, base_url: Constants.base_url()]
+  defp default_options,
+    do: %{
+      api: ConfigCat.API,
+      base_url: Constants.base_url_global(),
+      data_governance: DataGovernance.global(),
+      custom_endpoint: false,
+      redirects: %{}
+    }
+
+  defp custom_server_options(state, options) do
+    case List.keyfind(options, :base_url, 0) do
+      {:base_url, _url} ->
+        %{state | custom_endpoint: true}
+
+      _ ->
+        state
+    end
+  end
+
+  defp data_governance_options(state, options) do
+    with %{data_governance: data_governance, custom_endpoint: custom_endpoint} <-
+           Map.take(state, [:data_governance, :custom_endpoint]) do
+      cond do
+        custom_endpoint ->
+          {:base_url, base_url} = List.keyfind(options, :base_url, 0)
+          %{state | base_url: base_url}
+
+        data_governance == DataGovernance.eu_only() ->
+          %{state | base_url: Constants.base_url_eu_only()}
+
+        true ->
+          state
+      end
+    end
+  end
```

Comment:
I'm finding this option processing code quite hard to follow. Here are a few things that I'm having trouble with:

- It feels like the responsibilities of the various helpers could be separated a bit
- The mix of maps and keyword lists is a bit confusing
- `redirects` is in `default_options`, but isn't an option at all; it's only part of the GenServer state
- Having an argument named `state` in code that's processing options is confusing in a GenServer, which also has a `state`.

I'd like to propose the following refactoring, but it's up to you whether or not you want to take the suggestion. 

Here's what I did:
- Remove `base_url` from `default_options` entirely so that we can merge in the default options early, but still determine whether or not a `base_url` was provided by the caller.
- Process the options as a keyword list until we convert to a Map to become the initial_state
- Start with the same option processing code all of the other GenServers use (`default_options() |> Keyword.merge(options)`)
- Add a default base_url when necessary, and set the `custom_endpoint` flag at the same time
- Add additional state (`etag` and `redirects`) at the end

```elixir
  @spec start_link(options()) :: GenServer.on_start()
  def start_link(options) do
    {name, options} = Keyword.pop!(options, :name)

    initial_state =
      default_options()
      |> Keyword.merge(options)
      |> choose_base_url()
      |> Map.new()
      |> Map.merge(%{etag: nil, redirects: %{}})

    GenServer.start_link(__MODULE__, initial_state, name: name)
  end

  defp default_options,
    do: [api: ConfigCat.API, data_governance: DataGovernance.global()]

  defp choose_base_url(options) do
    case Keyword.get(options, :base_url) do
      nil ->
        base_url = options |> Keyword.get(:data_governance) |> default_url()
        Keyword.merge(options, base_url: base_url, custom_endpoint: false)

      _ ->
        Keyword.put(options, :custom_endpoint, true)
    end
  end

  defp default_url(DataGovernance.eu_only()), do: Constants.base_url_eu_only()
  defp default_url(_), do: Constants.base_url_global()
```

Reply:
Looks great! :)